### PR TITLE
Fix NCCL AllGather hanging issue for Qwen3 Next MTP

### PR DESCRIPTION
**Pull request by ispobock**
## Motivation

Fix hanging issue (AllGather size mismatch) in EAGLE speculative decoding (both V1 and V2) when running with TP>1 and non-greedy sampling. (https://github.com/sgl-project/sglang/issues/22276)

The actual root cause chain is:
same `next_token_logits` in non-greedy sampling branch
→ `softmax`/`top_k_renorm_prob`/`top_p_renorm_prob` produces slightly different `target_probs` across GPUs due to floating-point non-determinism 
→ `tree_speculative_sampling_target_only` samples different accepted token values 
→ divergent `output_ids` inserted into radix cache → different prefix match lengths on subsequent requests 
→ different `extend_seq_lens` and `num_seqs` → different AllGather sizes in LogitsProcessor → hang

## Modifications

 broadcast predict from rank 0 after the sampling kernel to ensure all TP ranks have identical accepted token values

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

File: python/sglang/srt/speculative/eagle_info.py [hunk 1]
```diff
@@ -377,6 +378,16 @@ def verify(
                 deterministic=True,
             )
 
+            # Sync sampling results across TP ranks: different GPUs may
+            # produce slightly different target_probs due to floating-point
+            # non-determinism in softmax/top_k/top_p, causing different
+            # sampled tokens. Broadcast from rank 0 to ensure consistency.
+            tp_group = get_tp_group()
```
Comment (yudian0504) [inline python/sglang/srt/speculative/eagle_info.py]: tp_group = get_attention_tp_group() if is_dp_attention_enabled() else get_tp_group()

File: python/sglang/srt/speculative/eagle_info.py [hunk 2]
```diff
@@ -377,6 +378,16 @@ def verify(
...
+            if tp_group.world_size > 1:
+                tp_group.broadcast(predict, src=0)
+                tp_group.broadcast(accept_index, src=0)
+                tp_group.broadcast(accept_length, src=0)
+
```
Comment (yudian0504) [inline python/sglang/srt/speculative/eagle_info.py:398]: Similarly, we should consider dp attn:
```python
dist.broadcast(predict, src=tp_group.ranks[0], group=tp_group.device_group)
dist.broadcast(accept_index, src=tp_group.ranks[0], group=tp_group.device_group)
dist.broadcast(accept_length, src=tp_group.ranks[0], group=tp_group.device_group)
```

File: python/sglang/srt/speculative/eagle_info.py (hunk 1 above)
Comment (ispobock) [reply python/sglang/srt/speculative/eagle_info.py]: Nice catch!

File: python/sglang/srt/speculative/eagle_info.py (hunk 2 above)
Comment (ispobock) [reply python/sglang/srt/speculative/eagle_info.py:398]: tp_group.broadcast already translates src=0 (group-local rank) to the correct global rank internally, so they're equivalent.

Comment (yudian0504) [reply python/sglang/srt/speculative/eagle_info.py:398]: You’re right, I just noticed that this wrapper already does the translations.

```python
    def broadcast(self, input_: torch.Tensor, src: int = 0):
        """Broadcast the input tensor.
        NOTE: `src` is the local rank of the source rank.
        """
        assert src < self.world_size, f"Invalid src rank ({src})"

        # Bypass the function if we are using only 1 GPU.
        if self.world_size == 1:
            return input_
        # Broadcast.
        torch.distributed.broadcast(
            input_, src=self.ranks[src], group=self.device_group
        )
        return input_
```

**Review by Qiaolin-Yu — APPROVED** [review]

**Comment by Han-xin58** [comment]
I have also encountered the same problem, please work hard to fix it as soon as possible
